### PR TITLE
Rework symlink creation to support K8s charms

### PR DESCRIPTION
**Pull request by dshcherb**
At the time of writing Juju handles kubernetes charms on CAAS models
differently from regular charms on IAAS models:

* install event is not executed for k8s charms (see LP: #1854635),
  although this is expected to change in future Juju versions;
* start event is executed first (unless there are storage events which
  run before that just like with the install event).

In order to support k8s charms, this change introduces the following:

* All event symlinks now point to ../lib/charm.py instead of pointing to
  the install symlink (this will also be useful for having a common code
  to set up functions as they are not placed into the hooks directory);
* Symlink creation mechanism is now triggered on 'start' event as well;

### CODE REVIEW

File: op/main.py [hunk 1]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
         # (otherwise, we would never have been called).
         return
 
-    event_hook_path = charm_dir / 'hooks' / bound_event.event_kind.replace('_', '-')
+    if issubclass(bound_event.event_type, op.charm.HookEvent):
+        event_dir = charm_dir / 'hooks'
+    else:
+        raise RuntimeError(f'unable to create a symlink: unsupported event type {bound_event.event_type}')
```
Comment (niemeyer) [inline op/main.py]: s/unable to/cannot/

File: op/main.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 import op.model
 
 CHARM_STATE_FILE = '.unit-state.db'
+CHARM_CODE_FILE = '../lib/charm.py'
```
Comment (niemeyer) [inline op/main.py]: That doesn't look right. Per agreements in the sprint, the charm.py file will be the entry point, and op/main.py inside a library. We cannot assume a relative path across them.

File: op/main.py [hunk 3]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
 
-    event_hook_path = charm_dir / 'hooks' / bound_event.event_kind.replace('_', '-')
+    if issubclass(bound_event.event_type, op.charm.HookEvent):
+        event_dir = charm_dir / 'hooks'
+    else:
+        raise RuntimeError(f'unable to create a symlink: unsupported event type {bound_event.event_type}')
+    # TODO: Handle function/action events here.
+
+    if not event_dir.exists():
+        raise RuntimeError(f'unable to create event symlink: {event_dir} directory does not exist')
```
Comment (niemeyer) [inline op/main.py]: s/unable to/cannot/

Why isn't it trying to create the directory, though?  If it's going as far as creating the symlink itself, don't see a good reason not to put the directory in place as well.

File: op/main.py [hunk 4]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
-    if event_hook_path.exists():
-        # Non-symlink entries and the ones not pointing to "install"
-        # need to be removed.
-        if not event_hook_path.is_symlink():
-            debugf(f'Hook entry at {event_hook_path} is not a symlink:'
-                   ' attempting to remove it.')
+    if event_path.exists():
+        # Non-symlink entries and the ones not pointing to the charm code file need to be removed.
+        if not event_path.is_symlink():
+            debugf(f'Event entry at {event_path} is not a symlink: attempting to remove it.')
```
Comment (niemeyer) [inline op/main.py]: s/Event entry/Path/

"Event entry" is not something people not reading this code can relate to.

File: op/main.py [hunk 5]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
             # May raise IsADirectoryError, e.g. in case it is a directory which
             # is unexpected and left to the developer or operator to handle.
-            event_hook_path.unlink()
-        elif os.readlink(event_hook_path) != 'install':
-            debugf(f'Removing entry {event_hook_path} as it does not point'
-                   ' to "install"')
-            event_hook_path.unlink()
+            event_path.unlink()
+        elif os.readlink(event_path) != CHARM_CODE_FILE:
```
Comment (niemeyer) [inline op/main.py]: That won't work as-is for the reason stated earlier.

File: op/main.py [hunk 6]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
+            debugf(f'Removing entry {event_path} as it does not point to {CHARM_CODE_FILE}')
```
Comment (niemeyer) [inline op/main.py]: s/entry/path/

File: op/main.py [hunk 7]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
+            event_path.unlink()
         else:
             create_link = False
 
     if create_link:
-        debugf(f'Creating a new relative symlink at {event_hook_path}'
-               f' to pointing to "install" located in {charm_dir}/hooks')
-        event_hook_path.symlink_to('install')
+        debugf(f'Creating a new relative symlink at {event_path} to pointing to {CHARM_CODE_FILE}')
```
Comment (niemeyer) [inline op/main.py]: s/to pointing/pointing/

File: op/main.py [hunk 8]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
+        elif os.readlink(event_path) != CHARM_CODE_FILE:
+            debugf(f'Removing entry {event_path} as it does not point to {CHARM_CODE_FILE}')
+            event_path.unlink()
         else:
             create_link = False
 
     if create_link:
-        debugf(f'Creating a new relative symlink at {event_hook_path}'
-               f' to pointing to "install" located in {charm_dir}/hooks')
-        event_hook_path.symlink_to('install')
```
Comment (niemeyer) [inline op/main.py]: Note that this worked fine without the issue mentioned.

File: op/main.py [hunk 9]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
+    """Set up links for supported events that originate from Juju.
 
     Whether a charm can handle an event or not can be determined by
     introspecting which events are defined on it.
 
-    Hooks are created as symlinks to "install" which may be either
-    a copy of main.py (which calls the main function) or a symlink
-    to it. Note that it is important to avoid creating a recursive
-    symlink which is why the install event itself is skipped.
+    Hooks or functions are created as symlinks to ../lib/charm.py.
```
Comment (niemeyer) [inline op/main.py]: Related to earlier points.

File: op/main.py [hunk 10]
```diff
@@ -148,6 +153,9 @@ def main(charm_class):
     # of events from debugging sessions.
     juju_event_name = Path(sys.argv[0]).name
 
+    if not juju_event_name:
+        raise RuntimeError('unable to determine an event name based on environment variables')
```
Comment (niemeyer) [inline op/main.py]: s/unable to/cannot/

File: op/main.py [hunk 11]
```diff
@@ -148,6 +153,9 @@ def main(charm_class):
     # of events from debugging sessions.
     juju_event_name = Path(sys.argv[0]).name
 
+    if not juju_event_name:
```
Comment (niemeyer) [inline op/main.py]: Empty line above may be dropped. These two lines are very tightly related: set and check.

File: op/main.py [hunk 12]
```diff
@@ -163,8 +171,9 @@ def main(charm_class):
         # When a charm is force-upgraded and a unit is in an error state Juju does not run upgrade-charm and
         # instead runs the failed hook followed by config-changed. Given the nature of force-upgrading
         # the hook setup code is not triggered on config-changed.
-        if (juju_event_name in ('install', 'upgrade-charm') or juju_event_name.endswith('-storage-attached')):
-            _setup_hooks(charm_dir, charm)
+        # 'start' event is included as Juju does not fire the install event for K8s charms (see LP: #1854635).
+        if (juju_event_name in ('install', 'start', 'upgrade-charm') or juju_event_name.endswith('-storage-attached')):
```
Comment (niemeyer) [inline op/main.py]: Was already there, but parenthesis seem unnecessary.

File: test/test_main.py [hunk 13]
```diff
@@ -100,12 +98,9 @@ def _clear_symlinks(cls):
...
             absolute_path = Path(r) / f
-            if absolute_path.name == 'install' and absolute_path.is_symlink():
-                if os.readlink(absolute_path) != cls.CHARM_PY_RELPATH:
-                    raise SymlinkTargetError(f'"{absolute_path.name}" link does not point to {cls.CHARM_PY_RELPATH}')
-            elif absolute_path.name.endswith('-storage-attached') and absolute_path.is_symlink():
-                if os.readlink(absolute_path) != 'install':
-                    raise SymlinkTargetError(f'"{absolute_path.name}" link does not point to "install"')
+            if absolute_path.name in ('install', 'start') or absolute_path.name.endswith('-storage-attached') and absolute_path.is_symlink():
+                if os.readlink(absolute_path) != op.main.CHARM_CODE_FILE:
+                    raise SymlinkTargetError(f'"{absolute_path.name}" link does not point to {op.main.CHARM_CODE_FILE}')
```
Comment (niemeyer) [inline test/test_main.py]: Will need reconsideration based on earlier conversation.

File: test/test_main.py [hunk 14]
```diff
@@ -261,21 +256,18 @@ def test_event_not_implemented(self):
...
 
-    def test_setup_hooks(self):
+    def test_setup_event_links(self):
         """Test auto-creation of symlinks for supported events.
         """
-        event_hooks = [f'hooks/{e.replace("_", "-")}'
-                       for e in charm.Charm.on.events().keys()
-                       if e != 'install']
-
+        event_hooks = [f'hooks/{e.replace("_", "-")}' for e in charm.Charm.on.events().keys()]
```
Comment (niemeyer) [inline test/test_main.py]: Is keys() required here?

File: test/test_main.py [hunk 15]
```diff
@@ -287,43 +279,47 @@ def _assess_setup_hooks(event_spec):
 
             for event_hook in event_hooks:
                 self.assertTrue(os.path.exists(event_hook))
-                self.assertEqual(os.readlink(event_hook), 'install')
-                self.assertEqual(os.readlink('hooks/install'),
-                                 self.CHARM_PY_RELPATH)
-
+                self.assertEqual(os.readlink(event_hook), op.main.CHARM_CODE_FILE)
```
Comment (niemeyer) [inline test/test_main.py]: Same here and in other places in tests.

File: op/main.py [hunk 16]
```diff
@@ -44,46 +45,50 @@ def _handle_event_link(charm_dir, bound_event):
...
+            event_path.unlink()
         else:
             create_link = False
 
     if create_link:
-        debugf(f'Creating a new relative symlink at {event_hook_path}'
-               f' to pointing to "install" located in {charm_dir}/hooks')
-        event_hook_path.symlink_to('install')
+        debugf(f'Creating a new relative symlink at {event_path} to pointing to {CHARM_CODE_FILE}')
+        event_path.symlink_to(CHARM_CODE_FILE)
```
Comment (johnsca) [inline op/main.py]: As mentioned on Telegram, my preference is to drop the hard-coded `CHARM_CODE_FILE` value in favor of `os.readlink(sys.argv[0])`, which trusts the charm author to have gotten the first link correct and just reuses that link target for all other symlinks.

File: op/main.py (hunk 3 above)
Comment (dshcherb) [reply op/main.py]: We expect at least an `install` or `start` symlink to be there to get a location of the charm code file from there (especially since, as you pointed out in https://github.com/canonical/operator/pull/63/files#r352619084, we shouldn't hardcode the relative symlink value in the `main.py` module itself).

Considering that, creating a directory like this only makes some sense for `functions`/`actions` which means we will need to special-case that.

Comment (dshcherb) [reply op/main.py]: I would leave this as-is until we get to adding `functions`.

File: op/main.py (hunk 8 above)
Comment (dshcherb) [reply op/main.py]: I am suggesting to remove that to have common code for function events and regular events.
```
hooks/install
functions/<xyz> -> install  # this is wrong
```

File: op/main.py (hunk 7 above)
Comment (dshcherb) [reply op/main.py]: +1, typo.

File: test/test_main.py (hunk 14 above)
Comment (dshcherb) [reply test/test_main.py]: It's a dict:

```
(Pdb++) type(charm.Charm.on.events())
<class 'dict'>
```

File: op/main.py (hunk 2 above)
Comment (dshcherb) [reply op/main.py]: In order to address this, I suggest this https://github.com/canonical/operator/pull/63/commits/5e99602d5281bec9b115fe765047755da8a28574

Please let me know if that looks right to you and, if so, I will add more test coverage.

File: op/main.py (hunk 5 above)
Comment (dshcherb) [reply op/main.py]: See https://github.com/canonical/operator/pull/63/files#r352704416

File: op/main.py (hunk 16 above)
Comment (dshcherb) [reply op/main.py]: Thanks for the suggestion, see https://github.com/canonical/operator/pull/63/files#r352704416

File: op/main.py (hunk 9 above)
Comment (dshcherb) [reply op/main.py]: See https://github.com/canonical/operator/pull/63/files#r352704416

File: test/test_main.py (hunk 13 above)
Comment (dshcherb) [reply test/test_main.py]: See https://github.com/canonical/operator/pull/63/files#r352704416

File: test/test_main.py (hunk 15 above)
Comment (dshcherb) [reply test/test_main.py]: See https://github.com/canonical/operator/pull/63/files#r352704416

File: op/main.py [hunk 17]
```diff
@@ -33,59 +33,92 @@ def _load_metadata(charm_dir):
...
+        charm_code_link = _try_link(start_entry)
+
+    if charm_code_link is not None:
+        charm_code_file = charm_dir / 'hooks' / charm_code_link
+        if not charm_code_file.exists():
+            raise RuntimeError(f'{charm_code_file} does not exist')
+        if not charm_code_file.is_file():
+            raise RuntimeError(f'{charm_code_file} is not a regular file')
+        return charm_code_link
+    raise RuntimeError(f'cannot determine a link to the charm code based on {charm_dir}/hooks/install and {charm_dir}/hooks/start entries')
```
Comment (johnsca) [inline op/main.py]: I think that all of this could be replaced with:

```python
charm_code_link = os.readlink(sys.argv[0])
```

If we're successfully executing this code, it already means that we've been started by a valid symlink, so we shouldn't really care whether that happens to be `install`, `start`, or one of the storage hooks (which are missing from this logic, currently).

File: op/main.py [hunk 18]
```diff
@@ -33,59 +33,92 @@ def _load_metadata(charm_dir):
...
-    if event_hook_path.exists():
-        # Non-symlink entries and the ones not pointing to "install"
-        # need to be removed.
-        if not event_hook_path.is_symlink():
-            debugf(f'Hook entry at {event_hook_path} is not a symlink:'
-                   ' attempting to remove it.')
+    if event_path.exists():
+        # Non-symlink entries and the ones not pointing to the charm code file need to be removed.
+        if not event_path.is_symlink():
+            debugf(f'Path at {event_path} is not a symlink: attempting to remove it.')
```
Comment (johnsca) [inline op/main.py]: Can you remind me why we want to do this?  If the charm author wants to put some non-symlink files in there, why wouldn't we assume that they know what they're doing and leave them alone?  Is this an attempt to preemptively handle upgrades from, say, a reactive charm to this framework?  If so, I wonder if we're adding complexity for a problem that we haven't fully thought through anyway and maybe we should just punt on it for now.

File: op/main.py [hunk 19]
```diff
@@ -163,8 +198,9 @@ def main(charm_class):
         # When a charm is force-upgraded and a unit is in an error state Juju does not run upgrade-charm and
         # instead runs the failed hook followed by config-changed. Given the nature of force-upgrading
         # the hook setup code is not triggered on config-changed.
-        if (juju_event_name in ('install', 'upgrade-charm') or juju_event_name.endswith('-storage-attached')):
-            _setup_hooks(charm_dir, charm)
+        # 'start' event is included as Juju does not fire the install event for K8s charms (see LP: #1854635).
+        if juju_event_name in ('install', 'start', 'upgrade-charm') or juju_event_name.endswith('-storage-attached'):
```
Comment (johnsca) [inline op/main.py]: Instead of trying to catch all of the cases where a hook other than `install` is the first thing to run, and potentially getting it wrong, I wonder if we couldn't use our persistence layer to have a flag as to whether we've done the initial setup and then do it when the flag is not set (which is guaranteed to be the first run, regardless of what the actual hook is) and on `upgrade-charm` (which is always going to be a special case, and we might want to handle slightly differently anyway at some point).

File: op/main.py (hunk 17 above)
Comment (dshcherb) [reply op/main.py]: Looks like we will have some portability issues when it comes to using symlinks and `argv[0]`.

What I noticed is with `Xenial+` (python 3.5+) `argv[0]` is indeed a symlink. However, with Windows (I tested on Windows Server 2016 and python 3.7.5) it is an absolute path even if the file is a [Windows symlink](https://docs.microsoft.com/en-us/windows/win32/fileio/symbolic-links).

The logic to populate `argv[0]` is located here in cpython:
https://github.com/python/cpython/blob/3.6/Python/sysmodule.c#L2161-L2259 (3.6)
https://github.com/python/cpython/blob/3.7/Python/pathconfig.c#L280-L393 (3.7)

The Windows-specific replacement logic:
https://github.com/python/cpython/blob/3.7/Python/pathconfig.c#L352-L360
```
#if defined(MS_WINDOWS)
        /* Replace the first element in argv with the full path. */
        wchar_t *ptemp;
        if (GetFullPathNameW(argv0,
                           Py_ARRAY_LENGTH(fullpath),
                           fullpath,
                           &ptemp)) {
            argv0 = fullpath;
        }
#endif
```
Test:
```
#!/usr/bin/env python3

import sys

if __name__ == '__main__':
    print(sys.argv[0])
    import pdb
    pdb.set_trace()
```

```
C:\Users\Administrator\Downloads\test\lib\charm.py
> c:\users\administrator\downloads\test\lib\charm.py(9)<module>()
-> print(os.readlink(sys.argv[0]))
(Pdb) sys.argv[0]
'C:\\Users\\Administrator\\Downloads\\test\\lib\\charm.py
```

Comment (dshcherb) [reply op/main.py]: Also, `pathlib` does not create valid symlinks with Linux-style paths (the target is empty) unless a `Path` object is specified or Windows-style slashes are used:

```
Python 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 14 2019, 23:09:19) [MSC v.1916 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pathlib
>>> sp = pathlib.Path('start')
>>> sp.symlink_to('../lib/charm.py')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 1337, in symlink_to
    self._accessor.symlink(target, self, target_is_directory)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: '../lib/charm.py' -> 'start'
>>> sp.symlink_to('../lib/charm.py')
>>> sp.resolve()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 1151, in resolve
    s = self._flavour.resolve(self, strict=strict)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 202, in resolve
    s = self._ext_to_normal(_getfinalpathname(s))
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'start'
```

`os.readlink` returns a Linux-style path from the Windows symlink. This is why Windows makes the target empty AFAIKS:
```
>>> import os
>>> os.readlink(sp)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: readlink() argument 1 must be str, not WindowsPath
>>> os.readlink(str(sp))
'../lib/charm.py'
```

If either `Path('../lib/charm.py')` or `symlink_to('..\lib\charm.py')` are used, a relative symlink gets created correctly (although Windows explorer shows it as absolute, using `os.readlink` or [this windows explorer extension](https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html shows that it is relative):
```
>>> sp.unlink()
>>> spw = pathlib.Path('start')
>>> spw.symlink_to('..\lib\charm.py')
>>> spw.resolve()
WindowsPath('C:/Users/Administrator/Downloads/test/lib/charm.py')
>>> os.readlink(str(spw))
'..\\lib\\charm.py'
>>> spw.unlink()

>>> spw.symlink_to(pathlib.Path('../lib/charm.py'))
>>> os.readlink(str(spw))
'..\\lib\\charm.py'
```

Comment (dshcherb) [reply op/main.py]: Something like this should work on Linux distributions and Windows:

```pathlib.Path(symlink_name).symlink_to(os.path.relpath(os.path.realpath(sys.argv[0]), '/tmp/test/hooks'))```

I validated that the charm store does not mangle symlinks in the zip archive (the `charm` tool that reactive uses copies a hook template for all necessary events but this is not a general charm store behavior).

The problem is the portability of symlinks extracted from zip archives because when a standard Windows tool is used to uncompress a zip charm archive all symlinks are just **regular files** that contain the paths that symlinks contained before compression.

Juju, however, uses Go's `archive/zip` and it looks like there is some logic to handle symlinks:

https://github.com/juju/charm/blob/a35e524ea871ac01a0522bd37b6267a45e233fb2/charmarchive.go#L335-L339
https://github.com/juju/utils/blob/master/zip/zip.go#L41-L112
https://github.com/juju/utils/blob/master/zip/zip.go#L160-L171 (`os.Symlink(symlinkTarget, targetPath)`)

We will need to check the actual behavior on a Windows unit but it might just work if I read it correctly.

Comment (dshcherb) [reply op/main.py]: My simulation in the test charm shows that we cannot rely on argv[0] to determine an event name in a cross-platform manner.

test/charms/test_main/lib/charm.py:
```
if __name__ == '__main__':
    import sys
    sys.argv[0] = os.path.realpath(sys.argv[0])
    main(Charm)
```

```
[1] > /home/ubuntu/src/juju/operator/test/charms/test_main/lib/op/main.py(158)main()
-> juju_event_name = Path(sys.argv[0]).name
(Pdb++) n
[1] > /home/ubuntu/src/operator/juju/operator/test/charms/test_main/lib/op/main.py(159)main()
-> if not juju_event_name:
(Pdb++) juju_event_name
'charm.py'
```

I updated the bug https://bugs.launchpad.net/juju/+bug/1854505 with this info as we need an environment variable to reliably determine the event name.

File: op/main.py (hunk 19 above)
Comment (dshcherb) [reply op/main.py]: I think the approach with the persistence layer goes against the general desire to have idempotent hook executions where we evaluate the environment in order to make decisions about what to do next without relying too much on persistent state. Performance-wise this doesn't look like a problem. So I would rather keep this logic in the code itself.

When it comes to `upgrade-charm`, we always have to have it here as there may be new endpoints, storage or core events added for which we will need to create new symlinks (until Juju implements a single entrypoint file feature).

File: op/main.py (hunk 18 above)
Comment (dshcherb) [reply op/main.py]: OK, we had a discussion with Gustavo during the standup and we can ignore extra files.

File: op/main.py (hunk 17 above)
Comment (dshcherb) [reply op/main.py]: https://github.com/canonical/operator/pull/63/commits/e4a4df9eb3658595034bc24a96d874c724b76ab7

**Comment by dshcherb** [comment]
For the record, and in continuation of the comment about symlinks on Windows https://github.com/canonical/operator/pull/63#discussion_r353145355: a zip file extractor in Juju based on `archive/zip` correctly handles relative symlinks on Windows. I extracted the relevant parts of it and built a binary for Windows to test: https://gist.github.com/dshcherb/49d63f1fd78e605883c4150900974d52

The only problem for Windows is, as mentioned, the usage of `argv[0]` for getting an event name as Python resolves the script name if it is a symlink on Windows and places an absolute path to the target file instead.

We can handle that when Juju gets `JUJU_HOOK_NAME` exposed to the charm process and I suggest ignoring this problem in this PR.

File: test/test_main.py [hunk 20]
```diff
@@ -261,21 +258,18 @@ def test_event_not_implemented(self):
...
         install_link_path = JUJU_CHARM_DIR / 'hooks/install'
 
         # The symlink is expected to be present in the source tree.
         self.assertTrue(install_link_path.exists())
         # It has to point to main.py in the lib directory of the charm.
         self.assertEqual(os.readlink(install_link_path), self.CHARM_PY_RELPATH)
 
-        def _assess_setup_hooks(event_spec):
+        def _assess_event_links(event_spec):
             event_hook = JUJU_CHARM_DIR / f'hooks/{event_spec.event_name}'
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply test/test_main.py]: Thanks, reworked this in https://github.com/canonical/operator/pull/63/commits/78a0f5475adb3cc6e9b19125ba9871cd5b92a061

File: op/main.py [hunk 21]
```diff
@@ -33,59 +33,62 @@ def _load_metadata(charm_dir):
...
 
 
-def _handle_event_link(charm_dir, bound_event):
+def _handle_event_link(charm_dir, event_dir, charm_code_link, bound_event):
     """Create a symlink for a particular event.
 
     charm_dir -- A root directory of the charm
     bound_event -- An event for which to create a symlink.
     """
+    # TODO: Handle function/action events here.
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply op/main.py]: I have some WIP code which simply does this:

```
     if issubclass(bound_event.event_type, op.charm.HookEvent):
         event_dir = charm_dir / 'hooks'
     elif issubclass(bound_event.event_type, op.charm.FunctionEvent):
         event_dir = Path(charm_dir / 'functions')
         if not event_dir.exists():
             event_dir = Path(charm_dir / 'actions')
```

In other words, by the time we have to create symlinks, meta will have already been parsed and `FunctionEvents` defined via `define_event` in `CharmBase`.

**Comment by dshcherb** [comment]
@jameinel 

These lines validate all symlinks after every simulated event:
https://github.com/canonical/operator/blob/431ca3c79c0130e7b546758a765ce80634734ae5/test/test_main.py#L284-L286

File: test/test_main.py [hunk 22]
```diff
@@ -287,43 +286,62 @@ def _assess_setup_hooks(event_spec):
...
-            _assess_setup_hooks(event_spec)
-
-        self._clear_symlinks()
+        test_cases = [(
+            [
+                # Assess 'install' first because upgrade-charm or other events cannot be handled before install creates symlinks for them.
+                EventSpec(InstallEvent, 'install', charm_config=charm_config),
+            ], [
+                EventSpec(StartEvent, 'start', charm_config=charm_config),
+                EventSpec(ConfigChangedEvent, 'config-changed', charm_config=charm_config),
```
Comment (niemeyer) [inline test/test_main.py]: Per our conversation, this test needs to be reworked. It became too convoluted and copy and pasted for us to understand what it's doing anymore.

File: op/main.py [hunk 23]
```diff
@@ -33,59 +33,51 @@ def _load_metadata(charm_dir):
...
-    symlink which is why the install event itself is skipped.
+    Hooks or functions are created as symlinks to the charm code file which is determined by inspecting
+    symlinks provided by the charm author at hooks/install or hooks/start.
 
     charm_dir -- A root directory of the charm.
     charm -- An instance of the Charm class.
     """
+    # CPython has different implementations for populating sys.argv[0] for Linux and Windows. For Windows
+    # it is always an absolute path (any symlinks are resolved) while for Linux it can be a relative path.
+    charm_code_link = os.path.relpath(os.path.realpath(sys.argv[0]), charm_dir / 'hooks')
```
Comment (niemeyer) [inline op/main.py]: We agreed to also rename this to charm_exec_path.

File: test/test_main.py [hunk 24]
```diff
@@ -261,69 +262,63 @@ def test_event_not_implemented(self):
...
-        events_to_assess = (
-            EventSpec(InstallEvent, 'install', charm_config=charm_config),
-            EventSpec(StartEvent, 'start', charm_config=charm_config),
+        initial_combinations = {
+            'hooks_iaas': (
+                # 'install' comes first because upgrade-charm or other events cannot be handled before install creates symlinks for them.
+                EventSpec(InstallEvent, 'install', charm_config=charm_config),
+            ),
+            'hooks_iaas_storage': (
+                # Storage hooks run before "install" so this case needs to be checked as well.
```
Comment (niemeyer) [inline test/test_main.py]: How is this case relevant for this test?

File: test/test_main.py [hunk 25]
```diff
@@ -261,69 +262,63 @@ def test_event_not_implemented(self):
...
+                EventSpec(InstallEvent, 'install', charm_config=charm_config),
+                EventSpec(StorageAttachedEvent, 'disks-storage-attached', charm_config=charm_config),
+            ),  # Similar to the above test cases for Kubernetes charms which do not get the install event executed (see LP: #1854635).
+            'hooks_caas': (
+                EventSpec(StartEvent, 'start', charm_config=charm_config),
```
Comment (niemeyer) [inline test/test_main.py]: Same.. per the conversation yesterday, this test is still confusing the idea of "this might happen" with "this is being tested". We need to make more clear what is being tested here, and I don't think we need to verify whether all potential combinations would work.

For example, some thing I can actually come up while brainstorming on this:

1. Is an event that is supposed to setup links actually doing that?
2. Is such an event idempotent if run multiple times?

What else?

If indeed the checks look that simple and objective, then this whole test ought to be simpler, more readable, and more objective than it currently looks.

File: test/test_main.py (hunk 24 above)
Comment (dshcherb) [reply test/test_main.py]: Fixed in https://github.com/canonical/operator/pull/63/commits/bf4278d0cd3048231be3027cdbe9cd62264fa6b8

File: test/test_main.py (hunk 25 above)
Comment (dshcherb) [reply test/test_main.py]: > I don't think we need to verify whether all potential combinations would work

What I had in mind:

All **valid combinations** of initial symlinks that a developer might include into the charm must result in a correct outcome: **all** event symlinks are present after **any** event execution with initial events executed first and other events executed afterwards.

We do not unlink anything in the hooks directory anymore so checking that all symlinks are present after events that are not initial (install, start, *-storage-attached) is not as important anymore. However, I would still do it for some events to avoid having a test that knows the internals and to make sure other events do not break the invariant.

Comment (dshcherb) [reply test/test_main.py]: Without other events: https://github.com/canonical/operator/pull/63/commits/cec63e32c51fe54251a9f42035a49babe41628e1

File: test/test_main.py [hunk 26]
```diff
@@ -91,23 +92,19 @@ class TestCharmEvents(CharmEvents):
...
-            if absolute_path.name == 'install' and absolute_path.is_symlink():
-                if os.readlink(absolute_path) != cls.CHARM_PY_RELPATH:
-                    raise SymlinkTargetError(f'"{absolute_path.name}" link does not point to {cls.CHARM_PY_RELPATH}')
-            elif absolute_path.name.endswith('-storage-attached') and absolute_path.is_symlink():
-                if os.readlink(absolute_path) != 'install':
-                    raise SymlinkTargetError(f'"{absolute_path.name}" link does not point to "install"')
-            else:
-                absolute_path.unlink()
+    def _prepare_initial_hooks(cls):
+        shutil.copytree(Path(__file__).parent / 'test_hooks', JUJU_CHARM_DIR / 'hooks', symlinks=True)
```
Comment (niemeyer) [inline test/test_main.py]: The new organization of test_hooks on that loose directory doesn't feel great. It's a loose directory with some symlinks pointing to a wrong place, which are copied over something else completely different which only makes sense for this one test.

How about dropping that test directory altogether, and simply creating the symlinks?

Comment (dshcherb) [reply test/test_main.py]: https://github.com/canonical/operator/pull/63/files#diff-3e85841b8cff61e2a613aa9b0c63a081R100-R113
